File: tup/tup-csharp/tars/TarsDisplayer.cs
/**
 * Tencent is pleased to support the open source community by making Tars available.
 *
 * Copyright (C) 2016THL A29 Limited, a Tencent company. All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use this file except 
 * in compliance with the License. You may obtain a copy of the License at
 *
 * https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software distributed 
 * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR 
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the 
 * specific language governing permissions and limitations under the License.
 */


using System;
using System.Text;
using System.Collections.Generic;
using System.Collections;
using Tup;

namespace Tup.Tars
{
    /**
     * 格式化输出tars结构的所有属性
     * 主要用于调试或打日志
     */
    public class TarsDisplayer
    {
        private StringBuilder sb;
        private int _level = 0;

        private void ps(string fieldName)
        {
            for (int i = 0; i < _level; ++i)
            {
                sb.Append('\t');
            }

            if (fieldName != null)
            {
                sb.Append(fieldName).Append(": ");
            }
        }

        public TarsDisplayer(StringBuilder sb, int level)
        {
            this.sb = sb;
            this._level = level;
        }

        public TarsDisplayer(StringBuilder sb)
        {
            this.sb = sb;
        }

        public TarsDisplayer Display(bool b, string fieldName)
        {
            ps(fieldName);
            sb.Append(b ? 'T' : 'F').Append('\n');
            return this;
        }

        public TarsDisplayer Display(byte n, string fieldName)
        {
            ps(fieldName);
            sb.Append(n).Append('\n');
            return this;
        }

        public TarsDisplayer Display(char n, string fieldName)
        {
            ps(fieldName);
            sb.Append(n).Append('\n');
            return this;
        }

        public TarsDisplayer Display(short n, string fieldName)
        {
            ps(fieldName);
            sb.Append(n).Append('\n');
            return this;
        }

        public TarsDisplayer Display(int n, string fieldName)
        {
            ps(fieldName);
            sb.Append(n).Append('\n');
            return this;
        }

        public TarsDisplayer Display(long n, string fieldName)
        {
            ps(fieldName);
            sb.Append(n).Append('\n');
            return this;
        }

        public TarsDisplayer Display(float n, string fieldName)
        {
            ps(fieldName);
            sb.Append(n).Append('\n');
            return this;
        }

        public TarsDisplayer Display(double n, string fieldName)
        {
            ps(fieldName);
            sb.Append(n).Append('\n');
            return this;
        }

        public TarsDisplayer Display(string s, string fieldName)
        {
            ps(fieldName);
            if (null == s)
            {
                sb.Append("null").Append('\n');
            }
            else
            {
                sb.Append(s).Append('\n');
            }

            return this;
        }

        public TarsDisplayer Display(byte[] v, string fieldName)
        {
            ps(fieldName);
            if (null == v)
            {
                sb.Append("null").Append('\n');
                return this;
            }
            if (v.Length == 0)
            {
                sb.Append(v.Length).Append(", []").Append('\n');
                return this;
            }
            sb.Append(v.Length).Append(", [").Append('\n');
            TarsDisplayer jd = new TarsDisplayer(sb, _level + 1);
            foreach (byte o in v)
            {
                jd.Display(o, null);
            }
            Display(']', null);
            return this;
        }

        public TarsDisplayer Display(char[] v, string fieldName)
        {
            ps(fieldName);
            if (null == v)
            {
                sb.Append("null").Append('\n');
                return this;
            }
            if (v.Length == 0)
            {
                sb.Append(v.Length).Append(", []").Append('\n');
                return this;
            }
            sb.Append(v.Length).Append(", [").Append('\n');
            TarsDisplayer jd = new TarsDisplayer(sb, _level + 1);
            foreach (char o in v)
            {
                jd.Display(o, null);
            }
            Display(']', null);
            return this;
        }

        public TarsDisplayer Display(short[] v, string fieldName)
        {
            ps(fieldName);
            if (null == v)
            {
                sb.Append("null").Append('\n');
                return this;
            }
            if (v.Length == 0)
            {
                sb.Append(v.Length).Append(", []").Append('\n');
                return this;
            }
            sb.Append(v.Length).Append(", [").Append('\n');
            TarsDisplayer jd = new TarsDisplayer(sb, _level + 1);
            foreach (short o in v)
            {
                jd.Display(o, null);
            }
            Display(']', null);
            return this;
        }

        public TarsDisplayer Display(int[] v, string fieldName)
        {
            ps(fieldName);
            if (null == v)
            {
                sb.Append("null").Append('\n');
                return this;
            }
            if (v.Length == 0)
            {
                sb.Append(v.Length).Append(", []").Append('\n');
                return this;
            }
            sb.Append(v.Length).Append(", [").Append('\n');
            TarsDisplayer jd = new TarsDisplayer(sb, _level + 1);
            foreach (int o in v)
                jd.Display(o, null);
            Display(']', null);
            return this;
        }

        public TarsDisplayer Display(long[] v, string fieldName)
        {
            ps(fieldName);
            if (null == v)
            {
                sb.Append("null").Append('\n');
                return this;
            }
            if (v.Length == 0)
            {
                sb.Append(v.Length).Append(", []").Append('\n');
                return this;
            }
            sb.Append(v.Length).Append(", [").Append('\n');
            TarsDisplayer jd = new TarsDisplayer(sb, _level + 1);
            foreach (long o in v)
            {
                jd.Display(o, null);
            }
            Display(']', null);
            return this;
        }

        public TarsDisplayer Display(float[] v, string fieldName)
        {
            ps(fieldName);
            if (null == v)
            {
                sb.Append("null").Append('\n');
                return this;
            }
            if (v.Length == 0)
            {
                sb.Append(v.Length).Append(", []").Append('\n');
                return this;
            }
            sb.Append(v.Length).Append(", [").Append('\n');
            TarsDisplayer jd = new TarsDisplayer(sb, _level + 1);
            foreach (float o in v)
            {
                jd.Display(o, null);
            }
            Display(']', null);
            return this;
        }

        public TarsDisplayer Display(double[] v, string fieldName)
        {
            ps(fieldName);
            if (null == v)
            {
                sb.Append("null").Append('\n');
                return this;
            }
            if (v.Length == 0)
            {
                sb.Append(v.Length).Append(", []").Append('\n');
                return this;
            }
            sb.Append(v.Length).Append(", [").Append('\n');
            TarsDisplayer jd = new TarsDisplayer(sb, _level + 1);
            foreach (double o in v)
            {
                jd.Display(o, null);
            }
            Display(']', null);
            return this;
        }

        public TarsDisplayer Display<K, V>(Dictionary<K, V> m, string fieldName)
        {
            ps(fieldName);
            if (null == m)
            {
                sb.Append("null").Append('\n');
                return this;
            }
            if (m.Count == 0)
            {
                sb.Append(m.Count).Append(", {}").Append('\n');
                return this;
            }
            sb.Append(m.Count).Append(", {").Append('\n');
            TarsDisplayer jd1 = new TarsDisplayer(sb, _level + 1);
            TarsDisplayer jd = new TarsDisplayer(sb, _level + 2);
            foreach (KeyValuePair<K, V> en in m)
            {
                jd1.Display('(', null);
                jd.Display(en.Key, null);
                jd.Display(en.Value, null);
                jd1.Display(')', null);
            }
            Display('}', null);
            return this;
        }

        public TarsDisplayer Display<T>(T[] v, string fieldName)
        {
            ps(fieldName);
            if (null == v)
            {
                sb.Append("null").Append('\n');
                return this;
            }
            if (v.Length == 0)
            {
                sb.Append(v.Length).Append(", []").Append('\n');
                return this;
            }
            sb.Append(v.Length).Append(", [").Append('\n');
            TarsDisplayer jd = new TarsDisplayer(sb, _level + 1);
            foreach (T o in v)
            {
                jd.Display(o, null);
            }
            Display(']', null);
            return this;
        }

        public TarsDisplayer Display<T>(List<T> v, string fieldName)
        {
            if (null == v)
            {
                ps(fieldName);
                sb.Append("null").Append('\n');
                return this;
            }
            else
            {
                for (int i = 0; i < v.Count; i++)
                {
                    Display(v[i], fieldName);
                }
                return this;
            }
        }

        ////@SuppressWarnings("unchecked")
        public TarsDisplayer Display<T>(T o, string fieldName)
        {
            object oObject = o;

            if (null == o)
            {
                sb.Append("null").Append('\n');
            }

            else if (o is byte)
            {
                Display(((byte)oObject), fieldName);
            }
            else if (o is bool)
            {
                Display(((bool)oObject), fieldName);
            }
            else if (o is short)
            {
                Display(((short)oObject), fieldName);
            }
            else if (o is int)
            {
                Display(((int)oObject), fieldName);
            }
            else if (o is long)
            {
                Display(((long)oObject), fieldName);
            }
            else if (o is float)
            {
                Display(((float)oObject), fieldName);
            }
            else if (o is Double)
            {
                Display(((Double)oObject), fieldName);
            }
            else if (o is string)
            {
                Display((string)oObject, fieldName);
            }
            else if (o is TarsStruct)
            {
                Display((TarsStruct)oObject, fieldName);
            }
            else if (o is byte[])
            {
                Display((byte[])oObject, fieldName);
            }
            else if (o is bool[])
            {
                Display((bool[])oObject, fieldName);
            }
            else if (o is short[])
            {
                Display((short[])oObject, fieldName);
            }
            else if (o is int[])
            {
                Display((int[])oObject, fieldName);
            }
            else if (o is long[])
            {
                Display((long[])oObject, fieldName);
            }
            else if (o is float[])
            {
                Display((float[])oObject, fieldName);
            }
            else if (o is double[])
            {
                Display((double[])oObject, fieldName);
            }
            else if (o.GetType().IsArray)
            {
                Display((Object[])oObject, fieldName);
            }
            else if (o is IList)
            {
                IList list = (IList)oObject;

                List<object> tmplist = new List<object>();
                foreach (object obj in list)
                {
                    tmplist.Add(obj);
                }
                Display(tmplist, fieldName);
            }
            else
            {
                throw new TarsEncodeException("write object error: unsupport type.");
            }

            return this;
        }
        public TarsDisplayer Display(TarsStruct v, string fieldName)
        {
            Display('{', fieldName);
            if (null == v)
            {
                sb.Append('\t').Append("null");
            }
            else
            {
                v.Display(sb, _level + 1);
            }

            Display('}', null);
            return this;
        }
    }

}

File: java/core/src/main/java/com/qq/tars/protocol/util/TarsDisplayer.java
/**
 * Tencent is pleased to support the open source community by making Tars available.
 *
 * Copyright (C) 2016 THL A29 Limited, a Tencent company. All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software distributed
 * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */

package com.qq.tars.protocol.util;

import java.util.Collection;
import java.util.List;
import java.util.Map;

import com.qq.tars.common.util.HexUtil;
import com.qq.tars.protocol.tars.TarsStructBase;
import com.qq.tars.protocol.tars.exc.TarsEncodeException;

public final class TarsDisplayer {

    private StringBuilder sb;
    private int _level = 0;

    private void ps(String fieldName) {
        for (int i = 0; i < _level; ++i)
            sb.append('\t');
        if (fieldName != null) sb.append(fieldName).append(": ");
    }

    public TarsDisplayer(StringBuilder sb, int level) {
        this.sb = sb;
        this._level = level;
    }

    public TarsDisplayer(StringBuilder sb) {
        this.sb = sb;
    }

    public TarsDisplayer display(boolean b, String fieldName) {
        ps(fieldName);
        sb.append(b ? 'T' : 'F').append('\n');
        return this;
    }

    public TarsDisplayer displaySimple(boolean b, boolean bSep) {
        sb.append(b ? 'T' : 'F');
        if (bSep) {
            sb.append("|");
        }
        return this;
    }

    public TarsDisplayer display(byte n, String fieldName) {
        ps(fieldName);
        sb.append(n).append('\n');
        return this;
    }

    public TarsDisplayer displaySimple(byte n, boolean bSep) {
        sb.append(n);
        if (bSep) {
            sb.append("|");
        }
        return this;
    }

    public TarsDisplayer display(char n, String fieldName) {
        ps(fieldName);
        sb.append(n).append('\n');
        return this;
    }

    public TarsDisplayer displaySimple(char n, boolean bSep) {
        sb.append(n);
        if (bSep) {
            sb.append("|");
        }
        return this;
    }

    public TarsDisplayer display(short n, String fieldName) {
        ps(fieldName);
        sb.append(n).append('\n');
        return this;
    }

    public TarsDisplayer displaySimple(short n, boolean bSep) {
        sb.append(n);
        if (bSep) {
            sb.append("|");
        }
        return this;
    }

    public TarsDisplayer display(int n, String fieldName) {
        ps(fieldName);
        sb.append(n).append('\n');
        return this;
    }

    public TarsDisplayer displaySimple(int n, boolean bSep) {
        sb.append(n);
        if (bSep) {
            sb.append("|");
        }
        return this;
    }

    public TarsDisplayer display(long n, String fieldName) {
        ps(fieldName);
        sb.append(n).append('\n');
        return this;
    }

    public TarsDisplayer displaySimple(long n, boolean bSep) {
        sb.append(n);
        if (bSep) {
            sb.append("|");
        }
        return this;
    }

    public TarsDisplayer display(float n, String fieldName) {
        ps(fieldName);
        sb.append(n).append('\n');
        return this;
    }

    public TarsDisplayer displaySimple(float n, boolean bSep) {
        sb.append(n);
        if (bSep) {
            sb.append("|");
        }
        return this;
    }

    public TarsDisplayer display(double n, String fieldName) {
        ps(fieldName);
        sb.append(n).append('\n');
        return this;
    }

    public TarsDisplayer displaySimple(double n, boolean bSep) {
        sb.append(n);
        if (bSep) {
            sb.append("|");
        }
        return this;
    }

    public TarsDisplayer display(String s, String fieldName) {
        ps(fieldName);
        if (null == s) {
            sb.append("null").append('\n');
        } else {
            sb.append(s).append('\n');
        }

        return this;
    }

    public TarsDisplayer displaySimple(String s, boolean bSep) {
        if (null == s) {
            sb.append("null");
        } else {
            sb.append(s);
        }
        if (bSep) {
            sb.append("|");
        }
        return this;
    }

    public TarsDisplayer display(byte[] v, String fieldName) {
        ps(fieldName);
        if (null == v) {
            sb.append("null").append('\n');
            return this;
        }
        if (v.length == 0) {
            sb.append(v.length).append(", []").append('\n');
            return this;
        }
        sb.append(v.length).append(", [").append('\n');
        TarsDisplayer jd = new TarsDisplayer(sb, _level + 1);
        for (byte o : v)
            jd.display(o, null);
        display(']', null);
        return this;
    }

    public TarsDisplayer displaySimple(byte[] v, boolean bSep) {
        if (null == v || v.length == 0) {
            if (bSep) {
                sb.append("|");
            }
            return this;
        }

        sb.append(HexUtil.bytes2HexStr(v));
        if (bSep) {
            sb.append("|");
        }
        return this;
    }

    public TarsDisplayer display(char[] v, String fieldName) {
        ps(fieldName);
        if (null == v) {
            sb.append("null").append('\n');
            return this;
        }
        if (v.length == 0) {
            sb.append(v.length).append(", []").append('\n');
            return this;
        }
        sb.append(v.length).append(", [").append('\n');
        TarsDisplayer jd = new TarsDisplayer(sb, _level + 1);
        for (char o : v)
            jd.display(o, null);
        display(']', null);
        return this;
    }

    public TarsDisplayer displaySimple(char[] v, boolean bSep) {
        if (null == v || v.length == 0) {
            if (bSep) {
                sb.append("|");
            }
            return this;
        }

        sb.append(new String(v));
        if (bSep) {
            sb.append("|");
        }
        return this;
    }

    public TarsDisplayer display(short[] v, String fieldName) {
        ps(fieldName);
        if (null == v) {
            sb.append("null").append('\n');
            return this;
        }
        if (v.length == 0) {
            sb.append(v.length).append(", []").append('\n');
            return this;
        }
        sb.append(v.length).append(", [").append('\n');
        TarsDisplayer jd = new TarsDisplayer(sb, _level + 1);
        for (short o : v)
            jd.display(o, null);
        display(']', null);
        return this;
    }

    public TarsDisplayer displaySimple(short[] v, boolean bSep) {
        if (null == v || v.length == 0) {
            sb.append("[]");
            if (bSep) {
                sb.append("|");
            }
            return this;
        }
        sb.append("[");
        TarsDisplayer jd = new TarsDisplayer(sb, _level + 1);
        for (int i = 0; i < v.length; i++) {
            short o = v[i];
            if (i != 0) {
                sb.append("|");
            }
            jd.displaySimple(o, false);
        }
        sb.append("]");
        if (bSep) {
            sb.append("|");
        }

        return this;
    }

    public TarsDisplayer display(int[] v, String fieldName) {
        ps(fieldName);
        if (null == v) {
            sb.append("null").append('\n');
            return this;
        }
        if (v.length == 0) {
            sb.append(v.length).append(", []").append('\n');
            return this;
        }
        sb.append(v.length).append(", [").append('\n');
        TarsDisplayer jd = new TarsDisplayer(sb, _level + 1);
        for (int o : v)
            jd.display(o, null);
        display(']', null);
        return this;
    }

    public TarsDisplayer displaySimple(int[] v, boolean bSep) {
        if (null == v || v.length == 0) {
            sb.append("[]");
            if (bSep) {
                sb.append("|");
            }
            return this;
        }
        sb.append("[");
        TarsDisplayer jd = new TarsDisplayer(sb, _level + 1);
        for (int i = 0; i < v.length; i++) {
            int o = v[i];
            if (i != 0) {
                sb.append("|");
            }
            jd.displaySimple(o, false);
        }
        sb.append("]");
        if (bSep) {
            sb.append("|");
        }
        return this;
    }

    public TarsDisplayer display(long[] v, String fieldName) {
        ps(fieldName);
        if (null == v) {
            sb.append("null").append('\n');
            return this;
        }
        if (v.length == 0) {
            sb.append(v.length).append(", []").append('\n');
            return this;
        }
        sb.append(v.length).append(", [").append('\n');
        TarsDisplayer jd = new TarsDisplayer(sb, _level + 1);
        for (long o : v)
            jd.display(o, null);
        display(']', null);
        return this;
    }

    public TarsDisplayer displaySimple(long[] v, boolean bSep) {
        if (null == v || v.length == 0) {
            sb.append("[]");
            if (bSep) {
                sb.append("|");
            }
            return this;
        }
        sb.append("[");
        TarsDisplayer jd = new TarsDisplayer(sb, _level + 1);
        for (int i = 0; i < v.length; i++) {
            long o = v[i];
            if (i != 0) {
                sb.append("|");
            }
            jd.displaySimple(o, false);
        }
        sb.append("]");
        if (bSep) {
            sb.append("|");
        }

        return this;
    }

    public TarsDisplayer display(float[] v, String fieldName) {
        ps(fieldName);
        if (null == v) {
            sb.append("null").append('\n');
            return this;
        }
        if (v.length == 0) {
            sb.append(v.length).append(", []").append('\n');
            return this;
        }
        sb.append(v.length).append(", [").append('\n');
        TarsDisplayer jd = new TarsDisplayer(sb, _level + 1);
        for (float o : v)
            jd.display(o, null);
        display(']', null);
        return this;
    }

    public TarsDisplayer displaySimple(float[] v, boolean bSep) {
        if (null == v || v.length == 0) {
            sb.append("[]");
            if (bSep) {
                sb.append("|");
            }
            return this;
        }
        sb.append("[");
        TarsDisplayer jd = new TarsDisplayer(sb, _level + 1);
        for (int i = 0; i < v.length; i++) {
            float o = v[i];
            if (i != 0) {
                sb.append("|");
            }
            jd.displaySimple(o, false);
        }
        sb.append("]");
        if (bSep) {
            sb.append("|");
        }
        return this;
    }

    public TarsDisplayer display(double[] v, String fieldName) {
        ps(fieldName);
        if (null == v) {
            sb.append("null").append('\n');
            return this;
        }
        if (v.length == 0) {
            sb.append(v.length).append(", []").append('\n');
            return this;
        }
        sb.append(v.length).append(", [").append('\n');
        TarsDisplayer jd = new TarsDisplayer(sb, _level + 1);
        for (double o : v)
            jd.display(o, null);
        display(']', null);
        return this;
    }

    public TarsDisplayer displaySimple(double[] v, boolean bSep) {
        if (null == v || v.length == 0) {
            sb.append("[]");
            if (bSep) {
                sb.append("|");
            }
            return this;
        }
        sb.append("[");
        TarsDisplayer jd = new TarsDisplayer(sb, _level + 1);
        for (int i = 0; i < v.length; i++) {
            double o = v[i];
            if (i != 0) {
                sb.append("|");
            }
            jd.displaySimple(o, false);
        }
        sb.append("[");
        if (bSep) {
            sb.append("|");
        }
        return this;
    }

    public <K, V> TarsDisplayer display(Map<K, V> m, String fieldName) {
        ps(fieldName);
        if (null == m) {
            sb.append("null").append('\n');
            return this;
        }
        if (m.isEmpty()) {
            sb.append(m.size()).append(", {}").append('\n');
            return this;
        }
        sb.append(m.size()).append(", {").append('\n');
        TarsDisplayer jd1 = new TarsDisplayer(sb, _level + 1);
        TarsDisplayer jd = new TarsDisplayer(sb, _level + 2);
        for (Map.Entry<K, V> en : m.entrySet()) {
            jd1.display('(', null);
            jd.display(en.getKey(), null);
            jd.display(en.getValue(), null);
            jd1.display(')', null);
        }
        display('}', null);
        return this;
    }

    public <K, V> TarsDisplayer displaySimple(Map<K, V> m, boolean bSep) {
        if (null == m || m.isEmpty()) {
            sb.append("{}");
            if (bSep) {
                sb.append("|");
            }
            return this;
        }

        sb.append("{");
        TarsDisplayer jd = new TarsDisplayer(sb, _level + 2);
        boolean first = true;
        for (Map.Entry<K, V> en : m.entrySet()) {
            if (!first) {
                sb.append(",");
            }
            jd.displaySimple(en.getKey(), true);
            jd.displaySimple(en.getValue(), false);
            first = false;
        }
        sb.append("}");
        if (bSep) {
            sb.append("|");
        }
        return this;
    }

    public <T> TarsDisplayer display(T[] v, String fieldName) {
        ps(fieldName);
        if (null == v) {
            sb.append("null").append('\n');
            return this;
        }
        if (v.length == 0) {
            sb.append(v.length).append(", []").append('\n');
            return this;
        }
        sb.append(v.length).append(", [").append('\n');
        TarsDisplayer jd = new TarsDisplayer(sb, _level + 1);
        for (T o : v)
            jd.display(o, null);
        display(']', null);
        return this;
    }

    public <T> TarsDisplayer displaySimple(T[] v, boolean bSep) {
        if (null == v || v.length == 0) {
            sb.append("[]");
            if (bSep) {
                sb.append("|");
            }
            return this;
        }

        sb.append("[");
        TarsDisplayer jd = new TarsDisplayer(sb, _level + 1);
        for (int i = 0; i < v.length; i++) {
            T o = v[i];
            if (i != 0) {
                sb.append("|");
            }
            jd.displaySimple(o, false);
        }
        sb.append("]");
        if (bSep) {
            sb.append("|");
        }
        return this;
    }

    public <T> TarsDisplayer display(Collection<T> v, String fieldName) {
        if (null == v) {
            ps(fieldName);
            sb.append("null").append('\t');
            return this;
        } else {
            return display(v.toArray(), fieldName);
        }
    }

    public <T> TarsDisplayer displaySimple(Collection<T> v, boolean bSep) {
        if (null == v) {
            sb.append("[]");
            if (bSep) {
                sb.append("|");
            }
            return this;
        } else {
            return displaySimple(v.toArray(), bSep);
        }
    }

    @SuppressWarnings({ "unchecked", "rawtypes" })
    public <T> TarsDisplayer display(T o, String fieldName) {
        if (null == o) {
            sb.append("null").append('\n');
        } else if (o instanceof Byte) {
            display(((Byte) o).byteValue(), fieldName);
        } else if (o instanceof Boolean) {
            display(((Boolean) o).booleanValue(), fieldName);
        } else if (o instanceof Short) {
            display(((Short) o).shortValue(), fieldName);
        } else if (o instanceof Integer) {
            display(((Integer) o).intValue(), fieldName);
        } else if (o instanceof Long) {
            display(((Long) o).longValue(), fieldName);
        } else if (o instanceof Float) {
            display(((Float) o).floatValue(), fieldName);
        } else if (o instanceof Double) {
            display(((Double) o).doubleValue(), fieldName);
        } else if (o instanceof String) {
            display((String) o, fieldName);
        } else if (o instanceof Map) {
            display((Map) o, fieldName);
        } else if (o instanceof List) {
            display((List) o, fieldName);
        } else if (o instanceof TarsStructBase) {
            display((TarsStructBase) o, fieldName);
        } else if (o instanceof byte[]) {
            display((byte[]) o, fieldName);
        } else if (o instanceof boolean[]) {
            display((boolean[]) o, fieldName);
        } else if (o instanceof short[]) {
            display((short[]) o, fieldName);
        } else if (o instanceof int[]) {
            display((int[]) o, fieldName);
        } else if (o instanceof long[]) {
            display((long[]) o, fieldName);
        } else if (o instanceof float[]) {
            display((float[]) o, fieldName);
        } else if (o instanceof double[]) {
            display((double[]) o, fieldName);
        } else if (o.getClass().isArray()) {
            display((Object[]) o, fieldName);
        } else {
            throw new TarsEncodeException("write object error: unsupport type.");
        }
        return this;
    }

    @SuppressWarnings({ "unchecked", "rawtypes" })
    public <T> TarsDisplayer displaySimple(T o, boolean bSep) {
        if (null == o) {
            sb.append("null").append('\n');
        } else if (o instanceof Byte) {
            displaySimple(((Byte) o).byteValue(), bSep);
        } else if (o instanceof Boolean) {
            displaySimple(((Boolean) o).booleanValue(), bSep);
        } else if (o instanceof Short) {
            displaySimple(((Short) o).shortValue(), bSep);
        } else if (o instanceof Integer) {
            displaySimple(((Integer) o).intValue(), bSep);
        } else if (o instanceof Long) {
            displaySimple(((Long) o).longValue(), bSep);
        } else if (o instanceof Float) {
            displaySimple(((Float) o).floatValue(), bSep);
        } else if (o instanceof Double) {
            displaySimple(((Double) o).doubleValue(), bSep);
        } else if (o instanceof String) {
            displaySimple((String) o, bSep);
        } else if (o instanceof Map) {
            displaySimple((Map) o, bSep);
        } else if (o instanceof List) {
            displaySimple((List) o, bSep);
        } else if (o instanceof TarsStructBase) {
            displaySimple((TarsStructBase) o, bSep);
        } else if (o instanceof byte[]) {
            displaySimple((byte[]) o, bSep);
        } else if (o instanceof boolean[]) {
            displaySimple((boolean[]) o, bSep);
        } else if (o instanceof short[]) {
            displaySimple((short[]) o, bSep);
        } else if (o instanceof int[]) {
            displaySimple((int[]) o, bSep);
        } else if (o instanceof long[]) {
            displaySimple((long[]) o, bSep);
        } else if (o instanceof float[]) {
            displaySimple((float[]) o, bSep);
        } else if (o instanceof double[]) {
            displaySimple((double[]) o, bSep);
        } else if (o.getClass().isArray()) {
            displaySimple((Object[]) o, bSep);
        } else {
            throw new TarsEncodeException("write object error: unsupport type.");
        }
        return this;
    }

    public TarsDisplayer display(TarsStructBase v, String fieldName) {
        display('{', fieldName);
        if (null == v) {
            sb.append('\t').append("null");
        } else {
            v.display(sb, _level + 1);
        }

        display('}', null);
        return this;
    }

    public TarsDisplayer displaySimple(TarsStructBase v, boolean bSep) {
        sb.append("{");
        if (null == v) {
            sb.append('\t').append("null");
        } else {
            v.displaySimple(sb, _level + 1);
        }
        sb.append("}");
        if (bSep) {
            sb.append("|");
        }
        return this;
    }
}
